talk/morelanguage/morestl.tex

```diff
@@ -25,6 +25,121 @@ \subsection{More STL}
+        std::string_view sv {"Some example"};
+        auto first = sv.substr(0,sv.find_first_of(" "));
+        constexpr std::string_view location {"Geneva,CH"};
+        std::string some_string {"foo bar"};
+        std::string_view sv_str(some_string);
+        char foo[3] = {'f','o','o'};
+        std::string_view carr(foo, std::size(foo));
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[17]{std::optional}
+  \begin{itemize}
+  \item Indicates the absence of a value with \texttt{std::nullopt}
+  \item Has value semantics (Copy, Move, Compare, stack alloc.)
+  \item Useful in place of pointers where value semantics are intuitive
+  \end{itemize}
+  \begin{exampleblock}{Code example}
+    \begin{cppcode*}{}
+      std::optional<Phone> parse_phone(string_view in) {
+        if (is_valid_phone(in)) {
+          return in;// equiv. to optional<Phone>(in);
+        }
+        return {}; // default constructs std::nullopt
+      }
+      ...
+      auto v = parse_phone(get_input());
+      if (v) {  // alternatively v.is_valid()
+        process_phone(v.value()); // *v is equivalent
+      }
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[17]{std::variant}
+  \begin{itemize}
+  \item Expresses intent that variable holds one of the types
+  \item Makes it easy to implement visitor pattern
+  \end{itemize}
+
+  \begin{exampleblock}{Code example}
+    \begin{cppcode*}{}
+      using option_t = std::variant<int,float,string>;
+      option_t opt; // default constructs first type
+      std::variant<std::monostate, NonDefault, int> myvar;
+      opt = 100;
+      int ival = std::get<int>(opt);//also std::get<0>(opt)
+      try {
+        float val = std::get<float>(opt) // will throw...
+      } catch (const std::bad_variant_access& ex) {...}
+
+      // Much easier with get_if; returns a ptr or null
+      if (const float* val = std::get_if<float>(&opt))
+          std::cout<<"option type is f64, value="<< val;
+    \end{cppcode*}
+  \end{exampleblock}
+
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[17]{Visiting with std::variant}
+    \begin{cppcode*}{}
+      using option_t = std::variant<int,float,string>;
+      struct Visitor {
+        void operator() (int i) { std::cout<< "i32:"<< i;}
+        void operator() (float f) { std::cout<< "f32:"<< f;}
+        void operator() (string s) { std::cout<< "s:"<< s;}
+      };
+      void print_opt(option_t opt) {
+        std::visit(Visitor{}, opt);
+      }
+      ...
+      print_opt(100); print_opt(3.14f); print_opt("example");
+    \end{cppcode*}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[17]{std::any}
+  \begin{itemize}
+  \item Type safe alternative to void*
+  \item Useful for passing arbitrary values across interfaces that may not know
+    all the types
+  \item \texttt{any\_cast} will only match concrete types, inheritance hierarchy
+    is irrelevant
```

Comment:
Should we mention (or use colours) the fact (surprising for some people) that you cast like:
```c++
int *i = std::any_cast<int>(&any);
// and NOT
int *i = std::any_cast<int*>(&any);
```

Reply:
let's not mention it for now and see whether it's perturbing students or not